File: leetcode/0384-shuffle-an-array/solution.cpp
class Solution {
public:
    vector<int> originalNums;
    vector<int> nums;
    
    Solution(vector<int>& nums) {
        originalNums = nums;
        this->nums = nums;
    }
    
    /** Resets the array to its original configuration and return it. */
    vector<int> reset() {
        this->nums = originalNums;
        return this->nums;
    }
    
    /** Returns a random shuffling of the array. */
    vector<int> shuffle() {
        int sz = nums.size();
        for (int i = 0; i < sz; i++) {
            int idx = rand() % sz;
            int tmp = nums[i];
            nums[i] = nums[idx];
            nums[idx] = tmp;
        }
        
        return nums;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */

File: leetcode/0384-shuffle-an-array/Solution.java
class Solution {
    private int[] originalNums;
    private int[] nums;
    
    public Solution(int[] nums) {
        originalNums = Arrays.copyOfRange(nums, 0, nums.length);
        this.nums = nums;
    }
    
    /** Resets the array to its original configuration and return it. */
    public int[] reset() {
        nums = Arrays.copyOfRange(originalNums, 0, originalNums.length);
        return nums;
    }
    
    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
        Random random = new Random();
        for (int i = 0; i < nums.length; i++) {
            int idx = random.nextInt(nums.length);
            int tmp = nums[i];
            nums[i] = nums[idx];
            nums[idx] = tmp;
        }
        
        return nums;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */